samcli/commands/pipeline/bootstrap/cli.py

```diff
@@ -0,0 +1,221 @@
+    pipeline_ip_range: Optional[str],
+    confirm_changeset: bool,
+    config_file: Optional[str],
+    config_env: Optional[str],
+) -> None:
+    """
+    implementation of `sam pipeline bootstrap` command
+    """
+    if not pipeline_user_arn:
+        pipeline_user_arn = _load_saved_pipeline_user()
+
+    if interactive:
+        guided_context = GuidedContext(
+            stage_name=stage_name,
+            pipeline_user_arn=pipeline_user_arn,
+            pipeline_execution_role_arn=pipeline_execution_role_arn,
+            cloudformation_execution_role_arn=cloudformation_execution_role_arn,
+            artifacts_bucket_arn=artifacts_bucket_arn,
+            create_ecr_repo=create_ecr_repo,
+            ecr_repo_arn=ecr_repo_arn,
+            pipeline_ip_range=pipeline_ip_range,
+        )
+        guided_context.run()
+        stage_name = guided_context.stage_name
+        pipeline_user_arn = guided_context.pipeline_user_arn
+        pipeline_execution_role_arn = guided_context.pipeline_execution_role_arn
+        pipeline_ip_range = guided_context.pipeline_ip_range
+        cloudformation_execution_role_arn = guided_context.cloudformation_execution_role_arn
+        artifacts_bucket_arn = guided_context.artifacts_bucket_arn
+        create_ecr_repo = guided_context.create_ecr_repo
+        ecr_repo_arn = guided_context.ecr_repo_arn
+
+    if not stage_name:
+        raise click.UsageError("Missing required parameter '--stage-name'")
+
+    stage: Stage = Stage(
+        name=stage_name,
+        aws_profile=profile,
+        aws_region=region,
+        pipeline_user_arn=pipeline_user_arn,
+        pipeline_execution_role_arn=pipeline_execution_role_arn,
+        pipeline_ip_range=pipeline_ip_range,
+        cloudformation_execution_role_arn=cloudformation_execution_role_arn,
+        artifacts_bucket_arn=artifacts_bucket_arn,
+        create_ecr_repo=create_ecr_repo,
+        ecr_repo_arn=ecr_repo_arn,
+    )
+
+    stage.bootstrap(confirm_changeset=confirm_changeset)
+
+    stage.print_resources_summary()
+
+    try:
+        stage.save_config(
+            config_dir=PIPELINE_CONFIG_DIR, filename=PIPELINE_CONFIG_FILENAME, cmd_names=_get_command_name()
+        )
+    except Exception:
+        # Swallow saving exceptions, if any, as the resources are already bootstrapped and the ARNs are already
+        # printed out in the screen.
+        pass
+
+
+def _load_saved_pipeline_user() -> Optional[str]:
+    samconfig: SamConfig = SamConfig(config_dir=PIPELINE_CONFIG_DIR, filename=PIPELINE_CONFIG_FILENAME)
+    if not samconfig.exists():
+        return None
+    config: Dict[str, str] = samconfig.get_all(cmd_names=_get_command_name(), section="parameters")
+    return config.get("pipeline_user")
+
+
+def _get_command_name() -> List[str]:
+    ctx = click.get_current_context()
+    cmd_names: List[str] = cast(List[str], get_cmd_names(ctx.info_name, ctx))  # ["pipeline", "bootstrap"]
+    return cmd_names
```

Comment:
what is this function's purpose, looks like it is only used in `samconfig.get_all(cmd_names=_get_command_name(), section="parameters")`

Reply:
this returns the name of the sub-section of the toml file where we save the created infrastructure

Reply:
what are the possible return values of this `_get_command_name()`?

Reply:
it is always ["pipeline", "bootstrap"] but I don't want to hardcode it to be consistent with the other places in code

Reply:
that makes sense. `_get_command_name()` is a generic function right? If so I think it is better to place in a more common file (also let's add an example to its docstring, I was confused about what values it would generate)